jobbergate-api/jobbergate_api/apps/job_submissions/models.py

```diff
@@ -98,3 +105,87 @@ def include_parent(cls, query: Select) -> Select:
+        if value is not None:
+            return datetime.fromtimestamp(value, tz=timezone.utc)
+        return value
+
+    def process_result_value(self, value: datetime | None, dialect: Dialect) -> int | None:
+        if value is not None:
+            return int(value.timestamp())
+        return value
+
+
+class JobSubmissionMetric(CommonMixin, Base):
+    """
+    Job submission metric table definition.
+
+    Attributes:
+        time: The time the metric was recorded.
+        job_submission_id: The id of the job submission this metric is for.
+        slurm_job_id: The id of the job in the slurm queue.
+        node_host: The node on which the metric was recorded.
+        step: The step for which the metric was recorded.
+        task: The task for which the metric was recorded.
+        cpu_frequency: The CPU frequency at the time.
+        cpu_time: The CPU time (system + user) consumed at the time.
+        cpu_utilization: The CPU utilization (% of available) consumed at the time.
+        gpu_memory: The GPU memory consumed at the time (in MB).
+        gpu_utilization: The GPU utilizaiton (% of availavble) consumed at the time.
+        page_faults: The number of page faults at the time.
+        memory_rss: The resident set size of memory consumed at the time (in MB).
+        memory_virtual: The virtual memory allocated at the time (in MB).
+        disc_read: The amount of data read from disk at the time (in MB).
+        disk_write: The amount of data written to disk at the time (in MB).
+    """
+
+    time: Mapped[int] = mapped_column(TimestampInt, nullable=False, index=True)
+    job_submission_id: Mapped[int] = mapped_column(
+        Integer,
+        ForeignKey("job_submissions.id", ondelete="CASCADE"),
+        nullable=False,
+    )
+    slurm_job_id: Mapped[int] = mapped_column(Integer, nullable=False)
+    node_host: Mapped[str] = mapped_column(String, nullable=False, index=True)
+    step: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
+    task: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
+    cpu_frequency: Mapped[float] = mapped_column(Float, nullable=False)
+    cpu_time: Mapped[float] = mapped_column(Float, nullable=False)
+    cpu_utilization: Mapped[float] = mapped_column(Float, nullable=False)
+    gpu_memory: Mapped[int] = mapped_column(Integer, nullable=False)
+    gpu_utilization: Mapped[float] = mapped_column(Float, nullable=False)
+    page_faults: Mapped[int] = mapped_column(Integer, nullable=False)
+    memory_rss: Mapped[int] = mapped_column(Integer, nullable=False)
+    memory_virtual: Mapped[int] = mapped_column(Integer, nullable=False)
+    disk_read: Mapped[int] = mapped_column(Integer, nullable=False)
+    disk_write: Mapped[int] = mapped_column(Integer, nullable=False)
+
+    __table_args__ = (
+        PrimaryKeyConstraint("time", "job_submission_id", "node_host", "step", "task"),
```

Comment:
nice. In this case I followed [SQLAlchemy's docs](https://docs.sqlalchemy.org/en/20/core/constraints.html#setting-up-constraints-when-using-the-declarative-orm-extension):
> To apply table-level constraint objects such as [ForeignKeyConstraint](https://docs.sqlalchemy.org/en/20/core/constraints.html#sqlalchemy.schema.ForeignKeyConstraint) to a table defined using Declarative, use the __table_args__ attribute,

Reply:
Since we have to use the `__table_args__` for the indexes, we should leave this as it is. Notable, though, that the sqlalchemy docs don't establish a clear, standard way of doing this:
![image](https://github.com/user-attachments/assets/e88948de-0cff-4006-a4de-801d1f8e3b66)

Reply:
yes, but that's not how the tables are being declared though